discovery/aws/ec2.go

```diff
@@ -330,6 +332,13 @@ func (d *EC2Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error
 
 						for _, ipv6addr := range eni.Ipv6Addresses {
 							ipv6addrs = append(ipv6addrs, *ipv6addr.Ipv6Address)
+							if *ipv6addr.IsPrimaryIpv6 {
+								// we might have to extend the slice with more than one element
+								for int64(len(primaryipv6addrs)) <= *eni.Attachment.DeviceIndex {
+									primaryipv6addrs = append(primaryipv6addrs, "")
```

Comment:
How high can `*eni.Attachment.DeviceIndex` go?

Could this be a map instead of a slice?

Reply:
The maximum depends on the actual EC2 instance. According to https://instances.vantage.sh  the highest number could be 80 when using `trn1n.32xlarge`.
But the number of the Primary IPv6 addresses is less than or equal to the size of the `__meta_ec2_ipv6_addresses` where we use a slice already. (Every Primary IPv6 is an IPv6 address but not every IPv6 address is a Primary IPv6.)
Do you want me to change to map? I think in this PR we shouldn't touch the already existing metrics to keep the complexity low.

Reply:
I think it would be worth putting a guard on the code, like
```
const maxDeviceIndex = 1024
if *eni.Attachment.DeviceIndex  > maxDeviceIndex {
  return Error(...)
}
```

I withdraw the suggestion to use a map, but a comment here about the desire to retain gaps as "" would be good.

Reply:
I added a comment as you requested.

I can add the guard too but 1024 is too big for reality (80 could be the max at the moment). Also this protects only the Primary IPv6 addresses list as the size can't be bigger than the maximum DeviceIndex. The IPv4 and IPv6 addresses can be more numerous. For example a simple `c7g.4xlarge` instances can have:

* Maximum 8 ENIs -> maximum 8 Primary IPv6 addresses
* Maximum 30 IPvX address per ENI -> maximum 240 IPv4 and IPv6 addresses

The `__meta_ec2_ipv6_addresses` is not even ordered so I think it would be better to check the actual size of the slice and not the DeviceIndex value. My question is that do you want me to add this to the code? I prefer to not touch already existing code paths to keep the PRs complexity low and handle that in a different PR.

Reply:
The DeviceIndex cannot be greater than the maximum number of network interfaces minus one (the DeviceIndex starts from zero). So it cannot be greater than 79 in case someone has enough money to afford the `trn1n.32xlarge` instance.
I created a small `t4g.small` instance with 3 network interfaces. I detached the last of them (DeviceIndex=2) and tried to attach that again with DeviceIndex=3 but it failed with `AttachmentLimitExceeded`. It succeeded with DeviceIndex=2.
```
[ec2-user@i-09fd32b04973c0de7 ~]$ aws ec2 describe-instances --instance-ids i-09fd32b04973c0de7 | jq '.Reservations[0].Instances[0].NetworkInterfaces[] | { "ENI": .NetworkInterfaceId, "DeviceIndex": .Attachment.DeviceIndex, "AttachmentId": .Attachment.AttachmentId }'
{
  "ENI": "eni-019197b6acfcf0dd3",
  "DeviceIndex": 1,
  "AttachmentId": "eni-attach-0be9e07d6fe70015b"
}
{
  "ENI": "eni-076b468329cc12226",
  "DeviceIndex": 2,
  "AttachmentId": "eni-attach-07ddccc8bdc7a3c4c"
}
{
  "ENI": "eni-0f29c52f56da93ec8",
  "DeviceIndex": 0,
  "AttachmentId": "eni-attach-08e894e40c9281afa"
}
[ec2-user@i-09fd32b04973c0de7 ~]$ aws ec2 detach-network-interface --attachment-id eni-attach-07ddccc8bdc7a3c4c
[ec2-user@i-09fd32b04973c0de7 ~]$ aws ec2 describe-instances --instance-ids i-09fd32b04973c0de7 | jq '.Reservations[0].Instances[0].NetworkInterfaces[] | { "ENI": .NetworkInterfaceId, "DeviceIndex": .Attachment.DeviceIndex, "AttachmentId": .Attachment.AttachmentId }'
{
  "ENI": "eni-019197b6acfcf0dd3",
  "DeviceIndex": 1,
  "AttachmentId": "eni-attach-0be9e07d6fe70015b"
}
{
  "ENI": "eni-0f29c52f56da93ec8",
  "DeviceIndex": 0,
  "AttachmentId": "eni-attach-08e894e40c9281afa"
}
[ec2-user@i-09fd32b04973c0de7 ~]$ aws ec2 attach-network-interface --device-index 3 --instance-id i-09fd32b04973c0de7 --network-interface-id eni-076b468329cc12226

An error occurred (AttachmentLimitExceeded) when calling the AttachNetworkInterface operation: Device index 3 exceeds the limit for t4g.small
[ec2-user@i-09fd32b04973c0de7 ~]$ aws ec2 attach-network-interface --device-index 2 --instance-id i-09fd32b04973c0de7 --network-interface-id eni-076b468329cc12226
{
    "AttachmentId": "eni-attach-0ea97696143c99dd2",
    "NetworkCardIndex": 0
}
[ec2-user@i-09fd32b04973c0de7 ~]$ aws ec2 describe-instances --instance-ids i-09fd32b04973c0de7 | jq '.Reservations[0].Instances[0].NetworkInterfaces[] | { "ENI": .NetworkInterfaceId, "DeviceIndex": .Attachment.DeviceIndex, "AttachmentId": .Attachment.AttachmentId }'
{
  "ENI": "eni-019197b6acfcf0dd3",
  "DeviceIndex": 1,
  "AttachmentId": "eni-attach-0be9e07d6fe70015b"
}
{
  "ENI": "eni-0f29c52f56da93ec8",
  "DeviceIndex": 0,
  "AttachmentId": "eni-attach-08e894e40c9281afa"
}
{
  "ENI": "eni-076b468329cc12226",
  "DeviceIndex": 2,
  "AttachmentId": "eni-attach-0ea97696143c99dd2"
}

```
You can see that the order of the interfaces in the response can change. This is why I use the DeviceIndex to, well, index the new label. It doesn't matter if someone detaches an interface and attaches a new one until they use a different ENI as a scraping target. It will remain in the same "position" in the label.

Reply:
I see, thanks for trying that out.
I didn't know one can/should specify the index when attaching an interface.
That will give them even better control over the order (when detaching and re-attaching an ENI whose IPv6 is used as a scrape address via relabeling, they can avoid having to update the relabeling)